master/buildbot/process/buildrequestmerger.py

```diff
@@ -0,0 +1,246 @@
+from twisted.internet import reactor
+from twisted.python import log
+import sqlalchemy as sa
+
+from buildbot import config
+from buildbot.util.lru import LRUCache
+
+
+class BuildRequestMerger(config.ReconfigurableServiceMixin, service.Service):
+    """
+    Wrapper around buildsets.addBuildset that does merging before
+    buildrequests hit the db.
+
+    This will only do merges within a same buildchain, and will never merge
+    the top level build.
+    """
+
+    # Basic list of properties that must match for a buildrequest to be merged
+    # BuilderConfigs can define additional properties (see _propertiesMatch)
+    BASE_MERGE_PROPERTIES = ['force_rebuild', 'force_chain_rebuild']
+
+    def __init__(self, master):
+        self.master = master
+        self.startbrid_cache = LRUCache(
+            miss_fn=self.__startbridCacheMissFn, max_size=20000)
+        self.properties_cache = LRUCache(
+            miss_fn=None,  # Cache contents are handled manually
+            max_size=20000)
+
+    def addBuildset(self,
+                    sourcestampsetid,
+                    reason,
+                    properties,
+                    triggeredbybrid=None,
+                    builderNames=None,
+                    external_idstring=None,
+                    _reactor=reactor):
+        """
+        ..seealso:: buildsets.addBuildset
+            For parameter details
+        """
+        start = time.time()
+
+        buildsetLog = {
+            'name': 'addBuildset',
+            'description':
+            'Log merges within a chain while adding new buildsets',
+            'sourcestampsetid': sourcestampsetid,
+            'builderNames': builderNames,
+        }
+
+        # For every builderName in this buildset, check which ones can be merged
+        breqsToMerge = {}
+        for builderName in builderNames:
+            builderMergeStart = time.time()
+
+            if triggeredbybrid is None:
+                # This is a top level build, so it cannot be merged
+                mergeBrid = None
+            else:
+                # If we are not the top level build, find this chain's top level build
+                startbrid = self.startbrid_cache.get(triggeredbybrid)
+
+                # And look for a builder that matches the configured properties
+                mergeProperties = self.master.botmaster.builders[
+                    builderName].config.mergeProperties
+                mergeBrid = self._getMergeBrid(startbrid, builderName,
+                                               properties, mergeProperties)
+
+                # If we found one, add it to our merge map
+                if mergeBrid:
+                    breqsToMerge[builderName] = mergeBrid
+
+            buildsetLog[builderName] = {
+                'elapsed': time.time() - builderMergeStart,
+                'mergeBrid': mergeBrid
+            }
+
+        buildsetLog['elapsed'] = time.time() - start
+
+        log.msg(json.dumps(buildsetLog))
```

Comment:
I got this error when running locally 

> raise TypeError(repr(o) + " is not JSON serializable")
> 	exceptions.TypeError: set(['z-builder1']) is not JSON serializable

Reply:
I guess that is coming from `builderNames`.

Tests indicate that it should be receiving a list, but maybe that is wrong :P

I'll do a `sort` in it, that will improve log readability and handle sets at the same time.